File: Archivist/src/main/java/com/zorroa/archivist/security/WebSecurityConfig.java
package com.zorroa.archivist.security;

import com.zorroa.archivist.ArchivistConfiguration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationEventPublisher;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
import org.springframework.security.config.annotation.method.configuration.EnableGlobalMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.session.SessionRegistry;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;

@Configuration
@EnableWebSecurity
@EnableGlobalMethodSecurity(prePostEnabled=true)
public class WebSecurityConfig extends WebSecurityConfigurerAdapter {

    private static final Logger logger = LoggerFactory.getLogger(WebSecurityConfig.class);

    @Override
    protected void configure(HttpSecurity http) throws Exception {
        http
            .addFilterBefore(hmacAuthenticationFilter(), UsernamePasswordAuthenticationFilter.class)
            .authorizeRequests()
                .antMatchers("/").permitAll()
                .antMatchers("/health/**").permitAll()
                .antMatchers("/api/v1/analyst/_register").permitAll()
                .antMatchers("/api/v1/analyst/_shutdown").permitAll()
                .antMatchers("/console/**").hasAuthority("user::admin")
                .anyRequest().authenticated()
            .and()
                .httpBasic()
            .and().headers().frameOptions().disable()
            .and()
                .sessionManagement()
                    .maximumSessions(10)
                    .sessionRegistry(sessionRegistry())
                .and()
            .and()
               .csrf().disable();

    }

    @Autowired
    public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
        auth
            .authenticationProvider(authenticationProvider())
            .authenticationProvider(hmacAuthenticationProvider())
            .authenticationProvider(backgroundTaskAuthenticationProvider())
            .authenticationProvider(internalAuthenticationProvider())
            .authenticationEventPublisher(authenticationEventPublisher());

        /**
         * If its a unit test we add our rubber stamp authenticator.
         */
        if (ArchivistConfiguration.unittest) {
            auth.authenticationProvider(new UnitTestAuthenticationProvider());
        }
    }

    @Bean
    public AuthenticationEventPublisher authenticationEventPublisher() {
        return new AuthenticationEventPublisher() {

            @Override
            public void publishAuthenticationSuccess(
                    Authentication authentication) {
            }

            @Override
            public void publishAuthenticationFailure(
                    AuthenticationException exception,
                    Authentication authentication) {
            }
        };
    }

    @Bean
    public HmacSecurityFilter hmacAuthenticationFilter() {
        HmacSecurityFilter filter = new HmacSecurityFilter();
        return filter;
    }

    @Bean
    public AuthenticationProvider authenticationProvider() {
        return new ZorroaAuthenticationProvider();
    }

    @Bean
    public AuthenticationProvider hmacAuthenticationProvider() {
        return new HmacAuthenticationProvider();
    }

    @Bean
    public AuthenticationProvider backgroundTaskAuthenticationProvider() {
        return new BackgroundTaskAuthenticationProvider();
    }

    @Bean
    public AuthenticationProvider internalAuthenticationProvider() {
        return new InternalAuthenticationProvider();
    }

    @Bean
    public SessionRegistry sessionRegistry() {
        return new JdbcSessionRegistry();
    }
}

File: ldap/src/main/kotlin/com/zorroa/ldap/WebSecurityConfig.kt
package com.zorroa.ldap

import com.zorroa.archivist.sdk.security.LdapUserDetailsPlugin
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.ldap.DefaultSpringSecurityContextSource
import org.springframework.security.ldap.authentication.BindAuthenticator
import org.springframework.security.ldap.authentication.LdapAuthenticationProvider
import org.springframework.security.ldap.search.FilterBasedLdapUserSearch

@Configuration
@EnableWebSecurity
@Order(Ordered.HIGHEST_PRECEDENCE + 10)
open class WebSecurityConfig : WebSecurityConfigurerAdapter() {

    @Autowired
    lateinit var properties: LdapProperties

    @Throws(Exception::class)
    override fun configure(http: HttpSecurity) {
        http.formLogin().disable()
    }

    @Autowired
    open fun configureGlobal(auth: AuthenticationManagerBuilder) {
        if (!properties.enabled) {
            return
        }
        logger.info("Adding LDAP authentication provider")
        auth.authenticationProvider(ldapAuthenticationProvider())
    }

    @Bean
    @Throws(Exception::class)
    open fun ldapAuthenticationProvider(): AuthenticationProvider {
        val userDetailsService = ldapUserDetailsService()
        val contextSource = DefaultSpringSecurityContextSource(properties.url)
        contextSource.setBase(properties.base)
        contextSource.afterPropertiesSet()
        val ldapUserSearch = FilterBasedLdapUserSearch("", properties.filter, contextSource)
        val bindAuthenticator = BindAuthenticator(contextSource)
        bindAuthenticator.setUserSearch(ldapUserSearch)
        val ldapAuthenticationProvider = LdapAuthenticationProvider(bindAuthenticator, userDetailsService)
        ldapAuthenticationProvider.setUserDetailsContextMapper(userDetailsService)
        return ldapAuthenticationProvider
    }

    /**
     * If no LdapUserDetailsPlugin plugin is defined, use this one
     * as a default.
     **/
    @Bean
    @ConditionalOnMissingBean
    open fun ldapUserDetailsPlugin(): LdapUserDetailsPlugin {
        return StandardLdapUserDetailsPlugin()
    }

    @Bean
    open fun ldapUserDetailsService(): LdapUserDetailsServiceImpl {
        return LdapUserDetailsServiceImpl()
    }

    companion object {
        private val logger = LoggerFactory.getLogger(WebSecurityConfig::class.java)
    }
}
